### PR TITLE
Fix #4965. Add support for luhn searching and hashing.

### LINKED ISSUE #4965: Add support for luhn searching and hashing
https://en.m.wikipedia.org/wiki/Luhn_algorithm

### PR DESCRIPTION
**Pull request by m1ghtym0**
Test: https://github.com/radare/radare2-regressions/pull/494

### CODE REVIEW

File: libr/hash/luhn.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include <r_types.h>
+#include <r_util.h>
+
+R_API ut64 r_hash_luhn(const ut8 *buf, ut64 len) {
+	ut64 sum = 0;
+	int parity = (len-1) % 2;
+	char curChar[2] = "\0"; 
+	for (int i = len; i > 0 ; i--) {
+		curChar[0] = buf[i-1];
+		int curDigit = atoi(curChar);
```
Comment (radare) [inline libr/hash/luhn.c]: to get the numeric value of a single char there's no need to use atoi() or a 2 char buffer

File: libr/hash/luhn.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/* Copyright (C) radare2 - 2007-2016 - pancake */
```
Comment (radare) [inline libr/hash/luhn.c]: this copyright holder is not correct. i didnt wrote this code in 2007 :P put your nick/name and just 2016 in there

File: libr/hash/luhn.c [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <r_types.h>
+#include <r_util.h>
+
+R_API ut64 r_hash_luhn(const ut8 *buf, ut64 len) {
+	ut64 sum = 0;
+	int parity = (len-1) % 2;
```
Comment (radare) [inline libr/hash/luhn.c]: missing spaces around -

File: libr/hash/luhn.c [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	char curChar[2] = "\0"; 
+	for (int i = len; i > 0 ; i--) {
+		curChar[0] = buf[i-1];
+		int curDigit = atoi(curChar);
+
+		if (parity == i % 2) {
+			curDigit = curDigit * 2;
```
Comment (radare) [inline libr/hash/luhn.c]: use *= 2

File: libr/hash/luhn.h [hunk 5]
```diff
@@ -0,0 +1,6 @@
+#ifndef _LUHN_H_
+#define _LUHN_H_
+
+R_API ut64 r_hash_luhn(const ut8 *buf, ut64 len);
```
Comment (radare) [inline libr/hash/luhn.h]: put this in libr/include/r_hash.h 

this is the wrong place to put include files

File: libr/hash/luhn.h [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#ifndef _LUHN_H_
```
Comment (radare) [inline libr/hash/luhn.h]: breaks standard include barrier naming rules. it shouldnt start by underline (read it somewhere in some good practices book some day)

**Comment by radare** [comment]
i have merged the r2r commit, so plz fix this and i'll merge this . thanks!

**Comment by radare** [comment]
Fixed pending stuff and merged. Thanks!

**Comment by m1ghtym0** [comment]
Was a little busy, thanks for the fixes!

**Pull request closed without merging**